java-client/session/src/main/java/io/deephaven/client/impl/ObjectService.java

```diff
@@ -6,11 +6,123 @@
+     * A server object that supports a bidirectional message stream.
+     */
+    interface Bidirectional extends ServerObject {
+        /**
+         * The type.
+         *
+         * @return the type
+         */
+        String type();
+
+        /**
+         * Opens a bidirectional message stream for {@code this}. The returned {@link DataAndExports} messages are
+         * managed separately than {@code this}.
+         *
+         * <p>
+         * This provides a generic stream feature for Deephaven instances to use to add arbitrary functionality. This
+         * API lets a client open a stream to a particular object on the server, to be mediated by a server side plugin.
+         * In theory this could effectively be used to "tunnel" a custom gRPC call, but in practice there are a few
+         * deliberate shortcomings that still make this possible, but not trivial.
+         *
+         * <p>
+         * Presently it is required that the server respond immediately, at least to acknowledge that the object was
+         * correctly contacted (as opposed to waiting for a pending ticket, or dealing with network lag, etc). This is a
+         * small (and possibly not required, but convenient) departure from a offering a gRPC stream (a server-streaming
+         * or bidi-streaming call need not send a message right away).
+         *
+         * @param stream the stream where the client will receive messages
+         * @return the stream where the client will send messages
+         * @see Session#messageStream(HasTypedTicket, MessageStream) for the lower-level interface
+         */
+        MessageStream<DataAndTypedTickets> messageStream(MessageStream<DataAndExports> stream);
+    }
+
+    /**
+     * The sending and receiving interface for {@link #messageStream(HasTypedTicket, MessageStream)}.
+     *
+     * @param <Message> the message type
+     */
+    interface MessageStream<Message> {
+        void onData(Message message);
+
+        void onClose();
+    }
+
     /**
-     * Fetch the object.
+     * Exports {@code typedTicket} to a client-managed fetchable server object.
      *
-     * @param ticket the ticket
+     * @param typedTicket the typed ticket
      * @return the future
+     * @see Session#export(HasTypedTicket)
+     */
+    CompletableFuture<? extends Fetchable> fetchable(HasTypedTicket typedTicket);
+
+    /**
+     * Exports {@code typedTicket} to a client-managed bidirectional server object.
+     *
+     * @param typedTicket the typed ticket
+     * @return the future
+     * @see Session#export(HasTypedTicket)
+     */
+    CompletableFuture<? extends Bidirectional> bidirectional(HasTypedTicket typedTicket);
+
+    /**
+     * The low-level interface for fetching data. See {@link #fetchable(HasTypedTicket)} for a higher-level interface.
+     *
+     * @param typedTicket the typed ticket
+     * @return the future
+     */
+    CompletableFuture<DataAndExports> fetch(HasTypedTicket typedTicket);
```

Comment:
Maybe we could have overloads that just take typed tickets? Why do we need to "have" a typed ticket? I almost want something that is close to the gRPC.

Reply:
I think we should leave that for larger java client abstraction redesigns if necessary.